mmpy_bot/plugins/base.py

```diff
@@ -23,22 +42,38 @@ class Plugin(ABC):
     way, you can implement multithreading or multiprocessing as desired.
     """
 
-    def __init__(self):
+    def __init__(
+        self,
+        help_trigger: bool = False,
+        help_trigger_bang: bool = False,
+        direct_help: bool = False,
+    ):
         self.driver: Optional[Driver] = None
-        self.message_listeners: Dict[
-            re.Pattern, Sequence[MessageFunction]
-        ] = defaultdict(list)
-        self.webhook_listeners: Dict[
-            re.Pattern, Sequence[WebHookFunction]
-        ] = defaultdict(list)
+        self.direct_help: bool = direct_help
+        self.call_function: Optional[Callable] = None
+
+        self.message_listeners: Dict[re.Pattern, List[MessageFunction]] = defaultdict(
+            list
+        )
+        self.webhook_listeners: Dict[re.Pattern, List[WebHookFunction]] = defaultdict(
+            list
+        )
 
         # We have to register the help function listeners at runtime to prevent the
         # Function object from being shared across different Plugins.
-        self.help = listen_to("^help$", needs_mention=True)(Plugin.help)
-        self.help = listen_to("^!help$")(self.help)
+        # This code is a bit hairy because the function signature of an
+        # instance is (message) not (self, message) causing failures later
+        if help_trigger:
+            self.help = listen_to("^help$", needs_mention=True)(Plugin.help)
+        if help_trigger_bang:
+            if not help_trigger:
+                self.help = listen_to("^!help$")(Plugin.help)
+            else:
+                self.help = listen_to("^!help$")(self.help)
```

Comment:
```suggestion
        help_func = Plugin.help if (help_trigger_bang and help_trigger) else self.help
        self.help = listen_to("^!help$", needs_mention=help_trigger)(help_func)
```

I have no idea what the distinction means here; just simplifying the logic

Reply:
Your suggestion doesn't implement the same logic as mine but this code needs a better solution either way.
These conditionals are mostly there to allow the user to not register `help` or `!help` for every plugin which makes the help output unnecessarily verbose. For backwards compatibility I kept both behind independent boolean flags.